### ISSUE TITLE
Is it possible to create a remote debugging session?

### ISSUE THREAD

**Issue by ceremcem** [31i73/atom-dbg-gdb#29]
Is it possible to create a remote debugging session? If so, where do we assign

* gdb executable path: `/usr/bin/arm-none-eabi-gdb`  in my case 
* gdbinit file path

**Comment by ProPuke** [comment]
Yep. If you save a config it will generate a `.atom-dbg.cson` file in the root of your project, which you can put things into, or you can write one from scratch as either `.atom-dbg.cson` or `.atom-dbg.json` (for cson or json format).

The parameters are listed at the end of the package readme/info, but an example may look something like: (indentation is important)

```cson
"remote debug":
    debugger: "dbg-gdb"
    gdb_executable: "/usr/bin/arm-none-eabi-gdb"
    gdb_commands: ["source path_to_gdbinit"]
```

You'd then push F5 and select "remote debug" to run that config. You can also include multiple commands rather than bundling in a gdbinit if more convenient:
```cson
gdb_commands: ["foo", "bar", "baz"]
```

There's also `gdb_arguments` if you want to pass custom arguments to gdb when launching, or `cwd` if you need to set a custom cwd (it defaults to the first directory of the project).

Configs are named, so you can have as many as you want. If you save a configuration with the gui it will non-destructively append it to the end of the file. When you save changes the additions will become available immediately via F5.

**Comment by ceremcem** [comment]
Thank you for your response. 

Since my `gdb-init` is shared between several debuggers (command-line debugger will stay as an option), going forward with `gdb-init` is more appropriate right now. 

I added following cson file to the project root: 

```
"remote debug":
    debugger: "dbg-gdb"
    gdb_executable: "/usr/bin/arm-none-eabi-gdb"
    gdb_commands: ["source ./mcu-debug/gdb-init"]
```

and pressed F5, selected "remote debug". Debugger navigation buttons appeared: 

![image](https://user-images.githubusercontent.com/6639874/34026843-c54eec16-e169-11e7-80b8-14bf0629a202.png)

...but nothing more. No errors, no result. Looking my gdb server, there is no evidence of connection: 

![image](https://user-images.githubusercontent.com/6639874/34026972-bc8715d0-e16a-11e7-8b81-7bac2d6bee9f.png)

White section is gdb-server output (left) when I connect with `arm-none-eabi-gdb` (right). Red section is gdb-server's stand-by output which stays still when I hit F5 and select "`remote debug`" option.

**Comment by ProPuke** [comment]
Hmm.. It should just be literally forwarding those commands on to gdb. So if they're the correct commands it should work (I don't have a lot of experience with remote debugging I'm afraid).

If you view package settings for gdb-dbg there's an option to log output to the developer console, turn that on and enable the dev console in atom and you'll see the communication. Hopefully that helps narrow the problem.

**Comment by ceremcem** [comment]
Possibly this is the problem: 

![image](https://user-images.githubusercontent.com/6639874/34027553-6fd0acc0-e16e-11e7-9c0a-db3d850ff3f2.png)

**Comment by ProPuke** [comment]
Oh, ouch! That's a bug!

Are you by any chance running an old version of Atom? Looks like I used a method added around Atom v1.19 (or somewhere around there) but didn't set the version requirements appropriately. Sorry. Updating atom _should_ make it work.

**Comment by ceremcem** [comment]
Before I update the atom, do you want to add appropriate error message which I can confirm?

Edit: I created a snapshot of my system, I may make the error message confirmation at any time.

**Comment by ceremcem** [comment]
I updated Atom and F5 correctly connected to gdb server. Here is the current error message: 

![image](https://user-images.githubusercontent.com/6639874/34028431-2496d67a-e174-11e7-8881-3e6662fec9a5.png)

(Edit: `console` gives nothing.)

**Comment by ProPuke** [comment]
Ahh, whoops. I need to add an option to disable the run command when remote debugging. This will be fixed in the next update which will hopefully be this weekend. Sorry about that.

I know what the error is from before, so no need for the snapshot. But thanks for the help!

Both will be fixed in the next version. Cheers!

**Comment by creationix** [comment]
How's the progress on this going?  I'm trying to remote debug and am getting the same "Don't know how to run..." error.

I'm trying to connect to an openocd server.  Here is my `.atom-dbg.cson` file:

```cson
"Base-Debug/stm32f042f4px-daplie.elf":
  debugger: "dbg-gdb"
  gdb_executable: "/usr/bin/arm-none-eabi-gdb"
  gdb_commands: [
    "target remote :3333"
    "set print asm-demangle on"
    "monitor arm semihosting enable"
    "load"
    "step"
  ]
  path: "Base-Debug/stm32f042f4px-daplie.elf"
  cwd: "."
```

**Comment by ceremcem** [comment]
@creationix You may want to take a look at [mcu-debug](https://github.com/aktos-io/mcu-debug), the command line debugging tool for STM microcontrollers untill this issue is solved.

**Comment by papyDoctor** [comment]
For remote debugging, here is the solution:
Open ~/.atom/packages/dbg-gdb/lib

At lines 181 - 204 we have:
> @sendCommand '-exec-run'
> ...
> @dbg.stop()

Remove or comment (with #) these lines
See also solution for Uncaught Callback must be a function (https://github.com/31i73/atom-dbg/issues/46)

It seems that the maintainer of the debug packages is out
